storage-transcribe-audio/functions/src/transcribe-audio.ts

```diff
@@ -0,0 +1,138 @@
+import { ObjectMetadata } from "firebase-functions/v1/storage";
+import * as ffmpeg from "fluent-ffmpeg";
+import * as util from "util";
+
+import * as logs from "./logs";
+import config from "./config";
+
+export async function transcribeAudio(
+  client: SpeechClient,
+  object: ObjectMetadata
+): Promise<void> {
+  const request = {
+    config: {
+      encoding: "LINEAR16" as const,
+      sampleRateHertz: 48000,
+      languageCode: config.languageCode,
+      model: config.model,
+    },
+    audio: {
+      uri: `gs://${object.bucket}/${object.name}`,
+    },
+  };
+
+  const response = await transcribeLongOrUnknownLengthAudio(client, request);
+  logs.receivedLongRunningRecognizeResponse(response);
+  const transcript = response.results?.map(
+    (result) => result?.alternatives?.[0].transcript
+  );
+
+  if (!isStringArray(transcript)) {
+    logs.undefinedResultsTranscript(transcript);
+  } else {
+    logs.logResponseTranscript(transcript);
+  }
+
+  // TODO(reao): Handle user configs
+  // TODO(reao): Add tests
+  // TODO(reao): Handle files of known short length
+  // TODO(reao): Transcode files into lossless format (without upsampling)
+  // TODO(reao): Handle video files
+}
+
+type TranscodeAudioResult = TranscodeAudioSuccess | TranscodeAudioFailure;
+interface TranscodeAudioSuccess {
+  sampleRateHertz: number;
+  warnings: WarningType[];
+  channelCount: number;
+}
+
+interface TranscodeAudioFailure {
+  warnings: WarningType[];
+  type: FailureType;
+}
+
+enum FailureType {
+  UNKNOWN = 0,
+  ZERO_STREAMS = 1,
+  NULL_SAMPLE_RATE = 2,
+  NULL_CHANNELS = 3,
+}
+
+enum WarningType {
+  UNKNOWN = 0,
+  MORE_THAN_ONE_STREAM = 1,
+}
+
+export async function transcodeAudio(
+  localCopyPath: string,
+  outputPath: string
+): Promise<TranscodeAudioResult> {
+  const probePromise = util.promisify<string, ffmpeg.FfprobeData>(
+    ffmpeg.ffprobe
+  );
+  const probeData: ffmpeg.FfprobeData = await probePromise(localCopyPath);
+  const warnings: WarningType[] = [];
+
+  logs.debug("probeData:", probeData);
+
+  const { streams } = probeData;
+
+  if (streams.length == 0) {
+    logs.unexpectedZeroStreamCount();
+    return {
+      type: FailureType.ZERO_STREAMS,
+      warnings,
+    };
+  }
+  // TODO(reao): support multiple channels
+  if (streams.length != 1) {
+    logs.unexpectedStreamCount(streams.length);
+    warnings.push(WarningType.MORE_THAN_ONE_STREAM);
+  }
+
+  if (streams[0].sample_rate == null) {
+    logs.corruptedFile();
+    return {
+      type: FailureType.NULL_SAMPLE_RATE,
+      warnings,
+    };
+  }
+
+  if (streams[0].channels == null) {
+    logs.corruptedFile();
+    return {
+      type: FailureType.NULL_CHANNELS,
+      warnings,
+    };
+  }
+
+  return {
+    sampleRateHertz: streams[0].sample_rate,
+    channelCount: streams[0].channels,
+    warnings,
+  };
+}
+
+// For files whose length we don't know, or whose length is 60 secs or more.
```

Comment:
How does the implementation differ if the length is >60s vs <60s?

Reply:
For <60s, we can call a different function that runs faster -- see ["transcribe short audio files"](https://cloud.google.com/speech-to-text/docs/sync-recognize) vs ["transcribe long audio files"](https://cloud.google.com/speech-to-text/docs/async-recognize)

Reply:
After discussion, we've agreed to use <60s for all use cases, to use the API's upload capabilities.

Reply:
Huh does this mean we can't transcribe any files >60s?

Reply:
Wondering what the tradeoff was and why we decided with this.

Reply:
Ah! Typo! We've agreed to use >60s for all use cases, to use the upload capabilities of the API 😅

Reply:
oh gotcha, cool.

Reply:
I'm still not sure why we can't use the sync endpoint for <60s, I'm assuming it's because we can't know if the file is <60s unless we download it? If that's the case, we can tell developers they have to attach a metadata field to objects uploaded containing the duration (otherwise, it's assumed that its >60s). But up to you if you feel the complexity is worth it.

Reply:
We aren't using the sync endpoint for <60s because we're using a feature that's only available for the >60s endpoint. The feature we're using supports uploading results directly from the API to a Cloud Storage bucket, without piping the results through the extension; Alex recommended we use it because it's more resistant to failures. That feature isn't available for <60s.

Downloading the file isn't an issue because we download the file to transcode it for ease of use, so that the user can just upload any kind of audio file.